### PR TITLE
Apollo upgrade

### PR DESCRIPTION
**Pull request by LeonmanRolls**
## Please check changes you’ve made in the last month in case of a bad merge by me 

Local global state is now handled by typePolicies and reactive variables, new features in Apollo 3. See [here](https://www.apollographql.com/blog/apollo-client/caching/dispatch-this-using-apollo-client-3-as-a-state-management-solution/) for a good overview.

Our type policies are in `src/apollo/typePolicies`

Our reactive vars are in `src/apollo/reactiveVars`

Previously we were using `@client` to make web3 calls. This is no longer allowed and `@client` should only be used for local global state. 

Web3 queries are automatically picked up in the apollo middleware now. See `src/apollo/apolloClient`. We have a new web3Link that handles web3 calls.

If something needs to change in response to a local variable changing, that is handled in `useReactiveVarLIsteners` hook.

I've also started to build up our unit tests, lots of common use cases are handled by them so if you want to add some tests they can provide a good starting point.

We should adopt a more defensive coding style as the address and network can change at any time, no part of the app can expect that any particular piece of data will be available. Much of the work in this refactor is related to this as before we would wait for all data to be available before rendering or refresh on a change.

### CODE REVIEW

**Comment by makoto** [comment]
## 1

If you land to the site while connecting to non supported network, it stays at "Loading Network"

Screenshot 2021-09-09 at 17 26 27

## 2

If you switch to Matic after landing on the page, it says "Matic".

Screenshot 2021-09-09 at 17 27 09

Aa Matic is not supported, it got graphql error when I click to "My Account". 

## 3

While accessing http://localhost:3000/name/argent.xyz/details, I got the following error

Screenshot 2021-09-09 at 17 31 56

## 4.

Toggling network does not seem to always trigger the change on network.

Screenshot 2021-09-09 at 17 34 37

File: cypress/integration/nameDetail.spec.js [hunk 1]
```diff
@@ -40,12 +40,13 @@ function refreshAndCheckText(url, textOrArrayOfText) {
 describe('Name detail view', () => {
   it('can see list of top level domains from [root]', () => {
     cy.visit(`${NAME_ROOT}/[root]/subdomains`)
-    cy.queryByTestId('eth', { timeout: 10000 }).should('exist')
-    cy.queryByTestId('reverse', { timeout: 10000 }).should('exist')
+    cy.queryByTestId('eth', { timeout: 30000 }).should('exist')
```
Comment (makoto) [inline cypress/integration/nameDetail.spec.js:43]: We have to wait for 30 seconds?

File: cypress/integration/nameDetail.spec.js [hunk 2]
```diff
@@ -294,9 +296,10 @@ describe('Name detail view', () => {
         .type('FOOOOOOOO{enter}')
         .getByPlaceholderText('FOOOOOOOO', { exact: false })
         .type('Bar', { force: true, delay: 0 })
-      waitUntilInputResolves('Save').then(() => {
-        cy.getByText('Save').click({ force: true })
-      })
+      // waitUntilInputResolves('Save').then(() => {
+      //   cy.getByText('Save').click({ force: true })
+      // })
```
Comment (makoto) [inline cypress/integration/nameDetail.spec.js:301]: Can you remove these comments?

File: cypress/integration/registryMigration.spec.js [hunk 3]
```diff
@@ -6,7 +6,7 @@ const ENABLED_COLOUR = 'rgb(83, 132, 254)'
 
 describe('Migrate a subdomain to new registry', () => {
   it('can visit an unmigrated name and migrate it', () => {
-    cy.visit(`${ROOT}/name/sub1.testing.eth`)
+    cy.visit(`${ROOT}/name/sub1.testing.eth/details`)
```
Comment (makoto) [inline cypress/integration/registryMigration.spec.js:9]: Is this added just to speed up testing speed, or did you change the behaviour of how routing works?

File: src/apollo/mutations/mutations.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  provider?.on('chainChanged', _chainId => {
+    console.log('chain changed: ', _chainId)
+    setNetworkIdLocalMutation(parseInt(_chainId))
+    // getNetworkMutation()
+  })
+
+  provider?.on('accountsChanged', async accounts => {
+    console.log('accounts changed')
+    // setAccountsLocalMutation(accounts)
```
Comment (makoto) [inline src/apollo/mutations/mutations.js:36]: This function now doesn't do anything. If not needed, can you take it out?

File: src/api/subDomainRegistrar/resolvers.js [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
-      cache.writeData({
-        data: {
-          subDomainState: []
-        }
-      })
+      // cache.writeData({
+      //   data: {
+      //     subDomainState: []
+      //   }
+      // })
```
Comment (makoto) [inline src/api/subDomainRegistrar/resolvers.js:17]: can you get rid of comments

File: src/api/subDomainRegistrar/resolvers.js [hunk 6]
```diff
@@ -43,7 +43,7 @@ const resolvers = {
               subDomainState: [...cachedNodes]
             }
 
-            cache.writeData({ data })
+            // cache.writeData({ data })
```
Comment (makoto) [inline src/api/subDomainRegistrar/resolvers.js:46]: can you get rid of comments

File: src/api/resolverUtils.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import { GET_TRANSACTION_HISTORY } from '../graphql/queries'
-import getClient from '../apolloClient'
+import getClient from '../apollo/apolloClient'
+import { transactionHistoryReactive } from '../apollo/reactiveVars'
 
 async function addTransaction({ txHash, txState }) {
   const client = getClient()
```
Comment (makoto) [inline src/api/resolverUtils.js:6]: Doesn't look like this variable is in use so probably you can get rid of it

File: src/api/resolverUtils.js [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import { GET_TRANSACTION_HISTORY } from '../graphql/queries'
-import getClient from '../apolloClient'
+import getClient from '../apollo/apolloClient'
```
Comment (makoto) [inline src/api/resolverUtils.js:2]: Seems not in use any more

**Comment by makoto** [comment]
Looks like many CI tests have been failing https://app.travis-ci.com/github/ensdomains/ens-app/jobs/536569396 Maybe these tests are a bit fragile?

File: cypress/integration/registryMigration.spec.js (hunk 3 above)
Comment (LeonmanRolls) [reply cypress/integration/registryMigration.spec.js:9]: We had some duplicate code, so I am now doing a redirect when landing on that page

File: src/apollo/mutations/mutations.js (hunk 4 above)
Comment (LeonmanRolls) [reply src/apollo/mutations/mutations.js:36]: > Looks like many CI tests have been failing https://app.travis-ci.com/github/ensdomains/ens-app/jobs/536569396 Maybe these tests are a bit fragile?

mmm that's weird they were passing 2 days ago, but yes in general i've been struggling with the cypress tests

**Comment by LeonmanRolls** [comment]
> Looks like many CI tests have been failing https://app.travis-ci.com/github/ensdomains/ens-app/jobs/536569396 Maybe these tests are a bit fragile?

mmm that's weird they were passing 2 days ago, but yes in general i've been struggling with the cypress tests

**Comment by LeonmanRolls** [comment]
Also can you tell me how you managed to connect with matic? I can't reproduce

File: cypress/integration/nameDetail.spec.js (hunk 1 above)
Comment (LeonmanRolls) [reply cypress/integration/nameDetail.spec.js:43]: I think the internet is slow on travis sometimes

File: src/hooks/useReactiveVarListeners.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+`
+
+export default () => {
+  const {
+    data: { accounts, networkId, isENSReady }
+  } = useQuery(REACT_VAR_LISTENERS)
+
+  const previousNetworkId = usePrevious(networkId)
+
+  useEffect(() => {
```
Comment (LeonmanRolls) [inline src/hooks/useReactiveVarListeners.js:25]: Refactor to be an ordinary web3 call

File: src/hooks/useReactiveVarListeners.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    }
+  }, [accounts, isENSReady])
+
+  useEffect(() => {
+    if (previousNetworkId !== networkId && previousNetworkId !== undefined) {
+      const client = getClient()
+
+      client
+        .refetchQueries({
+          include: ['getRegistrations', 'getRegistrationsById', 'singleName'],
```
Comment (LeonmanRolls) [inline src/hooks/useReactiveVarListeners.js:40]: See if we can move this back into the component

File: src/api/manager/getReverseRecord.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import { normalize } from 'eth-ens-namehash'
```
Comment (LeonmanRolls) [inline src/api/manager/getReverseRecord.js:1]: try to remove the sideEffects copy of this

File: src/api/manager/resolvers.js [hunk 12]
```diff
@@ -394,7 +453,8 @@ const resolvers = {
           registrant
         ] = await Promise.all(dataSources)
 
-        const { names } = cache.readQuery({ query: GET_ALL_NODES })
+        // const { names } = cache.readQuery({ query: GET_ALL_NODES })
```
Comment (LeonmanRolls) [inline src/api/manager/resolvers.js:456]: remove comment

File: src/api/manager/resolvers.js [hunk 13]
```diff
@@ -426,15 +486,16 @@ const resolvers = {
           names: [...names, detailedNode]
         }
 
-        cache.writeData({ data })
+        // cache.writeData({ data })
```
Comment (LeonmanRolls) [inline src/api/manager/resolvers.js:489]: remove comment

File: src/api/manager/resolvers.js [hunk 14]
```diff
@@ -607,14 +628,16 @@ const resolvers = {
       const addresses = keys.map(key =>
         ens.getAddr(name, key).then(addr => ({ key, value: addr }))
       )
-      return Promise.all(addresses)
+      const test = await Promise.all(addresses)
```
Comment (LeonmanRolls) [inline src/api/manager/resolvers.js:631]: return promise

File: src/apollo/providerHelpers.js [hunk 15]
```diff
@@ -0,0 +1 @@
+export const getAccounts = async provider => {}
```
Comment (jefflau) [inline src/apollo/providerHelpers.js:1]: what is this doing?

File: src/apolloClient.js [hunk 16]
```diff
@@ -1,61 +0,0 @@
-import { ApolloClient } from 'apollo-client'
```
Comment (jefflau) [inline src/apolloClient.js:1]: Is this moved entirely to `apollo/apolloClient` ?

File: src/components/Links/EtherScanLink.spec.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+  __esModule: true,
+  ...jest.requireActual('@apollo/client'),
+  useQuery: jest.fn()
+}))
+import { useQuery } from '@apollo/client'
+
+import EtherScanLink from './EtherScanLink'
+import { StaticRouter } from 'react-router-dom'
+
+describe('EtherScanLink', () => {
```
Comment (jefflau) [inline src/components/Links/EtherScanLink.spec.js:13]: Liking these component tests. Maybe write in the readme best practices for testing components so we can keep them at a similar standard, especially for newcomers

File: src/components/QueryAccount.js [hunk 18]
```diff
@@ -1,39 +1,39 @@
...
-            return this.props.children({
-              account: '0x0000000000000000000000000000000000000000'
-            })
-          }
-          const { web3: { accounts } = { accounts: [] } } = data
-          return this.props.children({ account: accounts[0] })
-        }}
-      </Query>
-    )
+    return <div>QueryAccount.js was here</div>
```
Comment (jefflau) [inline src/components/QueryAccount.js:13]: What's this?

File: src/components/SingleName/Name.js [hunk 19]
```diff
@@ -32,36 +35,60 @@ function isRegistrationOpen(available, parent, isDeedOwner) {
...
-    return domain.parentOwner.toLowerCase() === account.toLowerCase()
+const NAME_REGISTER_DATA_WRAPPER = gql`
+  query nameRegisterDataWrapper @client {
+    accounts
+    networkId
   }
-  return false
+`
+
+export const useRefreshComponent = () => {
```
Comment (jefflau) [inline src/components/SingleName/Name.js:59]: Is this what force refreshes when network has changed? I'm sure I'm the reason this is here (although you look like you've abstracted to a hook), but maybe there's a better way?

File: src/components/SingleName/NameDetails.js [hunk 20]
```diff
@@ -618,66 +619,16 @@ function NameDetails({
...
-    return (
-      <NameRegister
-        registrationOpen={registrationOpen}
-        domain={domain}
-        refetch={refetch}
-        refetchIsMigrated={refetchIsMigrated}
-        readOnly={isEmptyAddress(account)}
-      />
-    )
+    return <Redirect to={`${pathname}/register`} />
```
Comment (jefflau) [inline src/components/SingleName/NameDetails.js:622]: What are the Redirects for?

File: src/components/SingleName/NameRegister/CTA.js [hunk 21]
```diff
@@ -255,6 +255,17 @@ const CTA = ({
...
+
+  useEffect(() => {
+    return () => {
+      if (step === 'REVEAL_CONFIRMED') {
+        refetch()
+      }
+    }
+  }, [step])
+
+  console.log('refetchIsMigrated: ', refetchIsMigrated)
```
Comment (jefflau) [inline src/components/SingleName/NameRegister/CTA.js:267]: remove this

File: src/components/SingleName/ResolverAndRecords/RecordsItem.js [hunk 22]
```diff
@@ -1,7 +1,8 @@
 import React from 'react'
 import PropTypes from 'prop-types'
 import styled from '@emotion/styled/macro'
-import { Mutation } from 'react-apollo'
+//import { Mutation } from 'react-apollo'
```
Comment (jefflau) [inline src/components/SingleName/ResolverAndRecords/RecordsItem.js:4]: Remove

File: src/components/SingleName/ResolverAndRecords/RecordsItem.js [hunk 23]
```diff
@@ -226,202 +227,191 @@ const RecordItemEditable = ({
...
+              isValid={isValid}
+              newValue={newValue}
+              startUploading={startUploading}
+              stopUploading={stopUploading}
+              stopAuthorizing={stopAuthorizing}
+            />
+          </>
+        ) : uploading && !authorized ? (
+          <>
+            <IpfsLogin startAuthorizing={startAuthorizing} />
```
Comment (jefflau) [inline src/components/SingleName/ResolverAndRecords/RecordsItem.js:394]: We can probably remove this as I guess it's not used? But maybe leave it for a separate PR/issue

File: src/graphql/queries.js [hunk 24]
```diff
@@ -1,26 +1,33 @@
 import gql from 'graphql-tag'
 import { NodeFields, SubDomainStateFields } from './fragments'
 
+// export const GET_WEB3 = gql`
```
Comment (jefflau) [inline src/graphql/queries.js:4]: Remove if not necessary

File: src/setup.js [hunk 25]
```diff
@@ -0,0 +1,143 @@
+import StackdriverErrorReporter from 'stackdriver-errors-js'
```
Comment (jefflau) [inline src/setup.js:1]: Is this the file that does all the setup when a network changes and we re-run this when something does change?

Comment (jefflau) [reply src/setup.js:1]: Note: let's go through this part in more detail over a call

**Comment by makoto** [comment]
I am getting this when connected to Matic.

Screenshot 2021-09-23 at 21 42 28

File: src/apollo/sideEffects.spec.js [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+
+jest.mock('./mutations/ens', () => ({
+  __esModule: true,
+  default: jest.fn()
+}))
+import getENS from './mutations/ens'
+
+import { getReverseRecord } from './sideEffects'
+
+describe('getReverseRecord', () => {
```
Comment (makoto) [inline src/apollo/sideEffects.spec.js:14]: Should you write a test case when it does not return the default?

File: src/apollo/apolloClient.js [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+
+    return fromPromise(
+      resolvers.Mutation[operationName]?.apply(null, [null, variables]),
+      operation
+    )
+  })
+
+  const splitLink = split(
+    ({ operationName }) => {
+      console.log(
```
Comment (makoto) [inline src/apollo/apolloClient.js:84]: can you get rid of the console.og?

File: src/apollo/sideEffects.js [hunk 28]
```diff
@@ -0,0 +1,43 @@
+import { emptyAddress } from '../utils/utils'
+import getENS from './mutations/ens'
+import { normalize } from 'eth-ens-namehash'
+import { isENSReady } from './reactiveVars'
+
+export const getReverseRecord = async address => {
```
Comment (makoto) [inline src/apollo/sideEffects.js:6]: Why is `getReverseRecord` on sideEffects? afaik, it's not changing any state.

File: src/components/AddReverseRecord.js [hunk 29]
```diff
@@ -317,4 +339,36 @@ function AddReverseRecord({ account, currentAddress }) {
...
+        queryDeduplication: false
+      }
+    }
+  )
+
+  useEffect(() => {
+    refetch()
+  }, [account, currentAddress, networkId])
+
+  console.log({ data })
```
Comment (makoto) [inline src/components/AddReverseRecord.js:368]: Can you get rid of these console.logs?

File: src/components/AddReverseRecord.js [hunk 30]
```diff
@@ -317,4 +339,36 @@ function AddReverseRecord({ account, currentAddress }) {
...
+  console.log({ error })
+
+  return <div>hi there</div>
```
Comment (makoto) [inline src/components/AddReverseRecord.js:371]: What's this? Should be removed?

File: src/components/DomainItem/ChildDomainItem.js [hunk 31]
```diff
@@ -10,7 +10,8 @@ import Tooltip from '../Tooltip/Tooltip'
 import QuestionMark from '../Icons/QuestionMark'
 import { checkIsDecrypted, truncateUndecryptedName } from '../../api/labels'
 import ExpiryDate from './ExpiryDate'
-import { Mutation } from 'react-apollo'
+//import { Mutation } from 'react-apollo'
```
Comment (makoto) [inline src/components/DomainItem/ChildDomainItem.js:13]: Please delete

File: src/components/NetworkInformation/NetworkInformation.js [hunk 32]
```diff
@@ -111,70 +109,53 @@ const WaitingText = styled('span')`
...
-  }
-
-  if (loading) {
-    return (
-      <Waiting>
-        <WaitingText>Waiting for accounts</WaitingText> <Loader />
-      </Waiting>
-    )
-  }
+  // if (loading) {
```
Comment (makoto) [inline src/components/NetworkInformation/NetworkInformation.js:140]: do we not need this any more? If not, please delete.

File: src/components/SingleName/NameRegister/registerReducer.test.js [hunk 33]
```diff
@@ -1,30 +1,9 @@
 import {
-  getStates,
   hasReachedState,
   registerMachine,
   registerReducer
 } from './registerReducer'
 
-test('getStates returns the correct states in order', () => {
-  expect(
-    getStates(
```
Comment (makoto) [inline src/components/SingleName/NameRegister/registerReducer.test.js:10]: Don't we still use getStates? If so, why do you remove the test?

File: src/components/SingleName/SubDomains.js [hunk 34]
```diff
@@ -134,6 +134,7 @@ function SubDomains({
               )
 
               return (
+                // <div>subdomain web3</div>
```
Comment (makoto) [inline src/components/SingleName/SubDomains.js:137]: Delete

File: src/routes/SearchResults.js [hunk 35]
```diff
@@ -71,65 +70,92 @@ class Results extends React.Component {
...
+    return (
+      <>
+        <SearchErrors errors={errors} searchTerm={searchTerm} />
+        {/*<SubDomainResults searchTerm={searchTerm} />*/}
+      </>
+    )
+  } else if (errors.length > 0) {
+    return <SearchErrors errors={errors} searchTerm={searchTerm} />
+  }
+  if (parsed) {
```
Comment (makoto) [inline src/routes/SearchResults.js:146]: `else if`?

File: src/routes/SingleName.js [hunk 36]
```diff
@@ -1,78 +1,93 @@
...
 function SingleName({
   match: {
     params: { name: searchTerm }
   },
   location: { pathname }
 }) {
   useScrollTo(0)
-  document.body.style.zoom = window.innerWidth / window.outerWidth
+
+  //document.body.style.zoom = window.innerWidth / window.outerWidth
```
Comment (makoto) [inline src/routes/SingleName.js:28]: Were you supposed to comment out to put it back ora going to delete?

**Comment by makoto** [comment]
Sometimes `fetchPolicy: 'no-cache'` comes with `context.queryDeduplication: false` and other times it does not. When do you use one way or another?

```
      fetchPolicy: 'no-cache',
      nextFetchPolicy: 'no-cache',
      context: {
        queryDeduplication: false
      }
``

File: src/routes/Home.js [hunk 37]
```diff
@@ -286,137 +267,91 @@ const ReadOnly = styled('span')`
   margin-left: 1em;
 `
 
+export const HOME_DATA = gql`
```
Comment (makoto) [inline src/routes/Home.js:270]: Previously all gql was in on its own file like `queries.js`. Are we now start embedding at routes and component level?

File: src/setup.js (hunk 25 above)
Comment (makoto) [reply src/setup.js:1]: `stackdriver-errors-js` can be removed now that @LeonmanRolls added different log report library

File: src/setup.js [hunk 38]
```diff
@@ -0,0 +1,143 @@
...
+        'labels',
+        JSON.stringify({
+          ...labels,
+          ...JSON.parse(process.env.REACT_APP_LABELS)
+        })
+      )
+      return provider
+    }
+
+    // const safe = await safeInfo()
```
Comment (makoto) [inline src/setup.js:61]: Delete

File: src/setup.js [hunk 39]
```diff
@@ -0,0 +1,143 @@
...
+  isAppReadyReactive,
+  isReadOnlyReactive,
+  networkIdReactive,
+  networkReactive,
+  reverseRecordReactive,
+  subDomainFavouritesReactive,
+  web3ProviderReactive
+} from './apollo/reactiveVars'
+import { setup as setupAnalytics } from './utils/analytics'
+import { safeInfo } from './utils/safeApps'
```
Comment (makoto) [inline src/setup.js:20]: `safeInfo` seems no longer in use

File: src/api/package.json [hunk 40]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (makoto) [inline src/api/package.json:1]: Why do you need package.json under api?
